### ISSUE TITLE
race condition between systemctl start and systemctl daemon-reload

### ISSUE THREAD

**Issue by jsynacek** [systemd/systemd#5328]
Executing the following deadlocks after a while:
```
# while true; do { systemctl start sshd & } ; systemctl daemon-reload ; fg ; done
```

The service was taken from Fedora 25.
```
# systemctl cat sshd
# /usr/lib/systemd/system/sshd.service
[Unit]
Description=OpenSSH server daemon
Documentation=man:sshd(8) man:sshd_config(5)
After=network.target sshd-keygen.target
Wants=sshd-keygen.target

[Service]
Type=forking
PIDFile=/var/run/sshd.pid
EnvironmentFile=-/etc/sysconfig/sshd
ExecStart=/usr/sbin/sshd $OPTIONS
ExecReload=/bin/kill -HUP $MAINPID
KillMode=process
Restart=on-failure
RestartSec=42s

[Install]
WantedBy=multi-user.target
```
Reproducible on the current master (commit 3b07d037f316196f05cc3d2d2812a3e8401a4c06).

**Comment by jsynacek** [comment]
```
(gdb) bt full
#0  0x00007f30f992eba9 in ppoll () from /lib64/libc.so.6
No symbol table info available.
#1  0x00007f30f9f0eac3 in bus_poll (bus=0x55d84d40b040, need_more=false, timeout_usec=18446744073709551615) at src/libsystemd/sd-bus/sd-bus.c:2915
        p = {{fd = 3, events = 1, revents = 0}, {fd = 0, events = 0, revents = 0}}
        r = 0
        e = 1
        n = 1
        ts = {tv_sec = -1, tv_nsec = 139848328475808}
        m = 18446744073709551615
        __PRETTY_FUNCTION__ = "bus_poll"
#2  0x00007f30f9f0ebf8 in sd_bus_wait (bus=0x55d84d40b040, timeout_usec=18446744073709551615) at src/libsystemd/sd-bus/sd-bus.c:2936
        __PRETTY_FUNCTION__ = "sd_bus_wait"
#3  0x00007f30f9ef7d1e in bus_process_wait (bus=0x55d84d40b040) at src/shared/bus-unit-util.c:818
        r = 0
#4  0x00007f30f9ef8c6c in bus_wait_for_jobs (d=0x55d84d40b920, quiet=false, extra_args=0x7fffd393d3a0) at src/shared/bus-unit-util.c:969
        q = 0
        r = 0
        __PRETTY_FUNCTION__ = "bus_wait_for_jobs"
        __func__ = "bus_wait_for_jobs"
#5  0x000055d84c58aa3f in start_unit (argc=2, argv=0x7fffd393d5c0, userdata=0x0) at src/systemctl/systemctl.c:3170
        q = 17
        arg_count = 0
        extra_args = {0x0, 0x0, 0x0, 0x0}
        w = 0x55d84d40b920
        method = 0x55d84c59eea3 "StartUnit"
        mode = 0x55d84c59dda5 "replace"
        one_name = 0x0
        suffix = 0x0
        names = 0x55d84d40b900
        bus = 0x55d84d40b040
        wait_context = {match = 0x0, event = 0x0, unit_paths = 0x0, any_failed = false}
        name = 0x55d84d40b908
        r = 0
        __func__ = "start_unit"
        __PRETTY_FUNCTION__ = "start_unit"
#6  0x00007f30f9eaef58 in dispatch_verb (argc=3, argv=0x7fffd393d5b8, verbs=0x55d84c5a9c40 <verbs>, userdata=0x0) at src/basic/verbs.c:92
        verb = 0x55d84c5a9d40 <verbs+256>
        name = 0x7fffd393eefb "start"
        i = 8
        left = 2
        __PRETTY_FUNCTION__ = "dispatch_verb"
        __func__ = "dispatch_verb"
#7  0x000055d84c59ca6c in systemctl_main (argc=3, argv=0x7fffd393d5b8) at src/systemctl/systemctl.c:8169
        verbs = {{verb = 0x55d84c5a5635 "list-units", min_args = 4294967295, max_args = 4294967295, flags = 3, dispatch = 0x55d84c58171a <list_units>}, {verb = 0x55d84c5a5640 "list-unit-files", min_args = 4294967295, max_args = 4294967295, flags = 0, dispatch = 0x55d84c584065 <list_unit_files>}, {verb = 0x55d84c5a5650 "list-sockets", min_args = 4294967295, max_args = 4294967295, flags = 2, dispatch = 0x55d84c582290 <list_sockets>}, {verb = 0x55d84c5a565d "list-timers", min_args = 4294967295, max_args = 4294967295, flags = 2, dispatch = 0x55d84c58369c <list_timers>}, {verb = 0x55d84c5a5669 "list-jobs", min_args = 4294967295, max_args = 4294967295, flags = 2, dispatch = 0x55d84c5871b2 <list_jobs>}, {verb = 0x55d84c5a5673 "list-machines", min_args = 4294967295, max_args = 4294967295, flags = 2, dispatch = 0x55d84c5861ea <list_machines>}, {verb = 0x55d84c5a06fd "clear-jobs", min_args = 4294967295, max_args = 1, flags = 2, dispatch = 0x55d84c593e28 <trivial_method>}, {verb = 0x55d84c5a0708 "cancel", min_args = 4294967295, max_args = 4294967295, flags = 2, dispatch = 0x55d84c587593 <cancel_job>}, {verb = 0x55d84c59ee9d "start", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c589fd4 <start_unit>}, {verb = 0x55d84c59eead "stop", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c589fd4 <start_unit>}, {verb = 0x55d84c59eebb "condstop", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c589fd4 <start_unit>}, {verb = 0x55d84c59eec4 "reload", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c589fd4 <start_unit>}, {verb = 0x55d84c59eed6 "restart", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c589fd4 <start_unit>}, {verb = 0x55d84c59eeea "try-restart", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c589fd4 <start_unit>}, {verb = 0x55d84c59ef11 "reload-or-restart", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c589fd4 <start_unit>}, {verb = 0x55d84c59ef64 "reload-or-try-restart", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c589fd4 <start_unit>}, {verb = 0x55d84c59ef37 "try-reload-or-restart", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c589fd4 <start_unit>}, {verb = 0x55d84c59ef85 "force-reload", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c589fd4 <start_unit>}, {verb = 0x55d84c59ef7a "condreload", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c589fd4 <start_unit>}, {verb = 0x55d84c59ef05 "condrestart", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c589fd4 <start_unit>}, {verb = 0x55d84c59f2b3 "isolate", min_args = 2, max_args = 2, flags = 2, dispatch = 0x55d84c589fd4 <start_unit>}, {verb = 0x55d84c5a5681 "kill", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c58c2ed <kill_unit>}, {verb = 0x55d84c5a5686 "is-active", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c58c20e <check_unit_active>}, {verb = 0x55d84c5a5690 "check", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c58c20e <check_unit_active>}, {verb = 0x55d84c5a5696 "is-failed", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c58c281 <check_unit_failed>}, {verb = 0x55d84c5a031f "show", min_args = 4294967295, max_args = 4294967295, flags = 2, dispatch = 0x55d84c592a55 <show>}, {verb = 0x55d84c5a56a0 "cat", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c5931e4 <cat>}, {verb = 0x55d84c5a02fc "status", min_args = 4294967295, max_args = 4294967295, flags = 2, dispatch = 0x55d84c592a55 <show>}, {verb = 0x55d84c5a0355 "help", min_args = 4294967295, max_args = 4294967295, flags = 2, dispatch = 0x55d84c592a55 <show>}, {verb = 0x55d84c5a56a4 "daemon-reload", min_args = 4294967295, max_args = 1, flags = 2, dispatch = 0x55d84c593bb2 <daemon_reload>}, {verb = 0x55d84c5a06c1 "daemon-reexec", min_args = 4294967295, max_args = 1, flags = 2, dispatch = 0x55d84c593bb2 <daemon_reload>}, {verb = 0x55d84c5a56b2 "show-environment", min_args = 4294967295, max_args = 1, flags = 2, dispatch = 0x55d84c594321 <show_environment>}, {verb = 0x55d84c5a08ea "set-environment", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c594d10 <set_environment>}, {verb = 0x55d84c5a56c3 "unset-environment", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c594d10 <set_environment>}, {verb = 0x55d84c5a56d5 "import-environment", min_args = 4294967295, max_args = 4294967295, flags = 2, dispatch = 0x55d84c594f77 <import_environment>}, {verb = 0x55d84c59f1e2 "halt", min_args = 4294967295, max_args = 1, flags = 2, dispatch = 0x55d84c58bf6e <start_system_special>}, {verb = 0x55d84c59f20c "poweroff", min_args = 4294967295, max_args = 1, flags = 2, dispatch = 0x55d84c58bf6e <start_system_special>}, {verb = 0x55d84c59f223 "reboot", min_args = 4294967295, max_args = 2, flags = 2, dispatch = 0x55d84c58bf6e <start_system_special>}, {verb = 0x55d84c59f237 "kexec", min_args = 4294967295, max_args = 1, flags = 2, dispatch = 0x55d84c58bf6e <start_system_special>}, {verb = 0x55d84c59f25d "suspend", min_args = 4294967295, max_args = 1, flags = 2, dispatch = 0x55d84c58bf6e <start_system_special>}, {verb = 0x55d84c59f276 "hibernate", min_args = 4294967295, max_args = 1, flags = 2, dispatch = 0x55d84c58bf6e <start_system_special>}, {verb = 0x55d84c59f294 "hybrid-sleep", min_args = 4294967295, max_args = 1, flags = 2, dispatch = 0x55d84c58bf6e <start_system_special>}, {verb = 0x55d84c59f2fc "default", min_args = 4294967295, max_args = 1, flags = 2, dispatch = 0x55d84c58bc5e <start_special>}, {verb = 0x55d84c59f2da "rescue", min_args = 4294967295, max_args = 1, flags = 2, dispatch = 0x55d84c58bf6e <start_system_special>}, {verb = 0x55d84c59f2f2 "emergency", min_args = 4294967295, max_args = 1, flags = 2, dispatch = 0x55d84c58bf6e <start_system_special>}, {verb = 0x55d84c59f249 "exit", min_args = 4294967295, max_args = 2, flags = 2, dispatch = 0x55d84c58bc5e <start_special>}, {verb = 0x55d84c5a070f "reset-failed", min_args = 4294967295, max_args = 4294967295, flags = 2, dispatch = 0x55d84c5940a3 <reset_failed>}, {verb = 0x55d84c5a099d "enable", min_args = 2, max_args = 4294967295, flags = 0, dispatch = 0x55d84c5964fc <enable_unit>}, {verb = 0x55d84c5a09a4 "disable", min_args = 2, max_args = 4294967295, flags = 0, dispatch = 0x55d84c5964fc <enable_unit>}, {verb = 0x55d84c5a09ac "is-enabled", min_args = 2, max_args = 4294967295, flags = 0, dispatch = 0x55d84c5980ca <unit_is_enabled>}, {verb = 0x55d84c5a0bbd "reenable", min_args = 2, max_args = 4294967295, flags = 0, dispatch = 0x55d84c5964fc <enable_unit>}, {verb = 0x55d84c5a0bcb "preset", min_args = 2, max_args = 4294967295, flags = 0, dispatch = 0x55d84c5964fc <enable_unit>}, {verb = 0x55d84c5a56e8 "preset-all", min_args = 4294967295, max_args = 1, flags = 0, dispatch = 0x55d84c597a52 <preset_all>}, {verb = 0x55d84c5a0bd2 "mask", min_args = 2, max_args = 4294967295, flags = 0, dispatch = 0x55d84c5964fc <enable_unit>}, {verb = 0x55d84c5a0bd7 "unmask", min_args = 2, max_args = 4294967295, flags = 0, dispatch = 0x55d84c5964fc <enable_unit>}, {verb = 0x55d84c5a0bc6 "link", min_args = 2, max_args = 4294967295, flags = 0, dispatch = 0x55d84c5964fc <enable_unit>}, {verb = 0x55d84c5a0bde "revert", min_args = 2, max_args = 4294967295, flags = 0, dispatch = 0x55d84c5964fc <enable_unit>}, {verb = 0x55d84c5a56f3 "switch-root", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c59453f <switch_root>}, {verb = 0x55d84c5a56ff "list-dependencies", min_args = 4294967295, max_args = 2, flags = 2, dispatch = 0x55d84c58544f <list_dependencies>}, {verb = 0x55d84c5a5711 "set-default", min_args = 2, max_args = 2, flags = 0, dispatch = 0x55d84c5865fe <set_default>}, {verb = 0x55d84c5a571d "get-default", min_args = 4294967295, max_args = 1, flags = 0, dispatch = 0x55d84c586355 <get_default>}, {verb = 0x55d84c5a5729 "set-property", min_args = 3, max_args = 4294967295, flags = 2, dispatch = 0x55d84c59389c <set_property>}, {verb = 0x55d84c5a5736 "is-system-running", min_args = 4294967295, max_args = 1, flags = 0, dispatch = 0x55d84c598657 <is_system_running>}, {verb = 0x55d84c5a0f96 "add-wants", min_args = 3, max_args = 4294967295, flags = 0, dispatch = 0x55d84c597454 <add_dependency>}, {verb = 0x55d84c5a0fa0 "add-requires", min_args = 3, max_args = 4294967295, flags = 0, dispatch = 0x55d84c597454 <add_dependency>}, {verb = 0x55d84c5a5748 "edit", min_args = 2, max_args = 4294967295, flags = 2, dispatch = 0x55d84c599c61 <edit>}, {verb = 0x0, min_args = 0, max_args = 0, flags = 0, dispatch = 0x0}}
#8  0x000055d84c59d9f2 in main (argc=3, argv=0x7fffd393d5b8) at src/systemctl/systemctl.c:8440
        r = 1
        __func__ = "main"
        __PRETTY_FUNCTION__ = "main"
```

**Comment by dtardon** [comment]
This issue was fixed by #8803 quite a while ago (and the fix later reworked in #11121).
